### PR TITLE
Updates RangeSize error message and tests for IPv6.

### PR DESCRIPTION
**Pull request by danehans**
**What this PR does / why we need it**:
Updates the RangeSize function's error message and tests for IPv6. Converts RangeSize unit test to a table test and tests for success and failure cases. This is needed to support IPv6. Previously, it was unclear whether RangeSize supported IPv6 CIDRs. These updates make IPv6 support explicit.

**Which issue this PR fixes**
Partially fixes Issue #1443

**Special notes for your reviewer**:
/area ipv6

**Release note**:
```NONE
```

### CODE REVIEW

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (caesarxuchao) [inline pkg/registry/core/service/ipallocator/allocator.go]: Just curious, does everyone say /1 is "smaller" than /2? Will it be ambiguous?

File: pkg/registry/core/service/ipallocator/allocator_test.go
Comment (caesarxuchao) [inline pkg/registry/core/service/ipallocator/allocator_test.go]: Why can't we use the correct `addrs` and test addrs==RangeSize(cidr)?

**Comment by MrHohn** [comment]
Reassigning ipv6 PR.

/unassign

/assign @bowei

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (danehans) [inline pkg/registry/core/service/ipallocator/allocator.go]: That's a good point. Maybe shorter and longer is more appropriate for describing the "length" of the subnet mask. Let me know if you find these descriptions acceptable.

File: pkg/registry/core/service/ipallocator/allocator_test.go
Comment (danehans) [inline pkg/registry/core/service/ipallocator/allocator_test.go]: OK. I'll update.

**Comment by caesarxuchao** [comment]
lgtm, but will let someone from the networking team take a look.

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (pmichali) [inline pkg/registry/core/service/ipallocator/allocator.go]: Change looks fine. Is there a meaningful name that could be used for the "31" limit to make this a bit more understandable? MaxHosts? MaxHostBits?

Should this function return an error and be handled by caller, versus using panic? Or are there too many callers for it to change?

File: pkg/registry/core/service/ipallocator/allocator_test.go
Comment (pmichali) [inline pkg/registry/core/service/ipallocator/allocator_test.go]: Only concern is that only passing cases are covered.

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (pmichali) [inline pkg/registry/core/service/ipallocator/allocator.go]: Should the function signature be changed to return an error indication? This way, you could unit test for failure cases (I don't know how you can UT panics).

**Review by pmichali — COMMENTED** [review]
Some suggestions on the RangeSize function.

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (pmichali) [inline pkg/registry/core/service/ipallocator/allocator.go]: Alternatively, should the function just return the # networks, 0+, and then the caller can check for validity of the number of the range? Right now, it determines the range of hosts AND checks for exceeding a maximum, but not a minimum. Looks like some callers are doing checking.

File: pkg/registry/core/service/ipallocator/allocator_test.go
Comment (danehans) [inline pkg/registry/core/service/ipallocator/allocator_test.go]: Agreed. I had fail cases in my original commit but I then removed them. As you mention in your other comment, I think the func needs to be updated to return an error.

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (danehans) [inline pkg/registry/core/service/ipallocator/allocator.go]: I'm open to making the changes. Maybe it's best to implement the changes and the failure case UT's in a follow-on PR? Lets see what @MrHohn has to say.

Comment (pmichali) [inline pkg/registry/core/service/ipallocator/allocator.go]: My vote would be to address it now, since the submit queue is blocked. BTW the current assignment has @bowei, instead of MrHohn.

**Comment by danehans** [comment]
@bowei when you have time, do you mind taking a look at this PR?

**Comment by bowei** [comment]
taking a look

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (bowei) [inline pkg/registry/core/service/ipallocator/allocator.go]: If possible I would change this function to return an error, we should try to avoid panic() in long running server code.

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (danehans) [inline pkg/registry/core/service/ipallocator/allocator.go]: In my experience, subnet masks are expressed using "length". I prefer using shorter or longer with a / notation to describe a subnet mask.

Comment (pmichali) [inline pkg/registry/core/service/ipallocator/allocator.go]: Sure.  I'm just pointing out (the subtle point) that the method is returning the number of hosts available for the subnet, and wondering if the error message that would get returned, should focus on that, rather than on mask size. For example:
```
err: = fmt.Sprintf("too many hosts allowed for subnet (%v). Must be /%d or longer", subnet, bits-30)
```
However, see my last comment on L254, where maybe this should be redesigned to check for valid subnet lengths, rather than returning the number of hosts and having the caller doing some of the checking, with the rest done here. Then, the error messages could be in terms of mask size (from tests for it being too short or too long).

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (pmichali) [inline pkg/registry/core/service/ipallocator/allocator.go]: Looking at the usage of this, the function panics (should be returning error, IMHO), if the subnet is too large, and if not, returns the number of hosts available, and the caller then checks if the subnet is too small. Maybe this should be changed to CheckSubnetSize() and return error if too large or too small, and caller can just check return (error) for non-nil?

File: pkg/registry/core/rest/storage_core.go
Comment (bowei) [inline pkg/registry/core/rest/storage_core.go]: is there a reason why this ServiceClusterIPAllocator is Capitalized?

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (bowei) [inline pkg/registry/core/service/ipallocator/allocator.go]: shouldn't this be return &r, nil?

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (bowei) [inline pkg/registry/core/service/ipallocator/allocator.go]: `"subnet mask must be >= /2 for IPv4 or /98 for IPv6"` (change message to be about valid values)

File: pkg/registry/core/rest/storage_core.go
Comment (danehans) [inline pkg/registry/core/rest/storage_core.go]: It should not be capitalized. I updated to lowercase.

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (danehans) [inline pkg/registry/core/service/ipallocator/allocator.go]: Thanks for catching this. I have updated the return.

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (danehans) [inline pkg/registry/core/service/ipallocator/allocator.go]: I have updated the error message.

**Review by danehans — COMMENTED** [review]
Updated based on @bowei comments.

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (bowei) [inline pkg/registry/core/service/ipallocator/allocator.go]: Put exported error at the top:
```
ErrMaskTooSmall = error.New("subnet mask must be longer than /1 for IPv4 or /97 for IPv6")
```
then
```
return 0, ErrMaskTooSmall
```

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (bowei) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: A bit more future proof:
```
if err != nil {
  if err == ipallocator.ErrMaskTooSmall {
    allErrs = append(allErrs, field.Invalid(fldPath, subnet, "subnet mask is too short"))
  } else {
        allErrs = append(allErrs, field.Invalid(fldPath, subnet, "invalid service subnet"))
  }
}
```

**Review by bowei — COMMENTED** [review]
otherwise looks ok to me

**Comment by danehans** [comment]
@bowei sorry it took me so long to address your feedback. I have updated the PR based on your feedback and tests pass. Let me know if you need anything for a lgtm. Thanks for your help.

**Review by pmichali — APPROVED** [review]
/lgtm

**Comment by pmichali** [comment]
@bowei Can you re apply the approved label, as it wanted lgtm and once done it now wants the aproved again. Thanks!

**Comment by bowei** [comment]
@mikedanese approval?

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (rpothier) [inline pkg/registry/core/service/ipallocator/allocator.go]: should this be updated for ipv6? bits - ones can be larger than 31 for ipv6

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (leblancd) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: This method (ValidateIPNetFromString()) is used to test both the service CIDR and the cluster's pod CIDR (see lines 250 and 252 below). This means that the cluster's pod CIDR is subject to the same size restriction as the service CIDR. Looks like RangeSize() limits IPv6 CIDRs to have subnet masks that must be longer than that of a /96 subnet. I think we need to loosen this restriction for pod CIDRs (maybe allow cluster pod CIDR of /48 or thereabouts if posible?)

**Review by leblancd — CHANGES_REQUESTED** [review]

File: pkg/registry/core/service/ipallocator/allocator_test.go
Comment (leblancd) [inline pkg/registry/core/service/ipallocator/allocator_test.go]: Comparing this test case ("failure IPv4 /24 cidr") to the previous, I'm not sure this test case adds much value. In the previous test case, you check that RangeSize returns the expected 256. Now in this test case, you pass exactly the same CIDR as before, and compare the result (which you know is going to be 256) with 512, which you know is going to fail (i.e no additional test coverage).

**Review by leblancd — CHANGES_REQUESTED** [review]
Looks pretty good. Suggesting some minor changes in the UT.

File: pkg/registry/core/service/ipallocator/allocator_test.go
Comment (leblancd) [inline pkg/registry/core/service/ipallocator/allocator_test.go]: In this test case, since tc.addrs == 0 and tc.expect == false, there's no check that RangeSize actually returns an error. So for example, if RangeSize is possibly changed in the future to not error for this CIDR, this test case would still pass. How about modifying the test like the following to try a mid-sized cidr, a max-sized cidr, and an oversized cidr (checking the error string for the latter):
```
func TestRangeSize(t *testing.T) {
	testCases := []struct {
		name   string
		cidr   string
		addrs  int64
		expErr string
	}{
		{
			name:  "success IPv4 /24 cidr",
			cidr:  "192.168.1.0/24",
			addrs: 256,
		},
		{
			name:  "success IPv4 /2 cidr",
			cidr:  "192.168.1.0/2",
			addrs: 1073741824,
		},
		{
			name:   "failure unsupported IPv4 cidr",
			cidr:   "192.168.1.0/1",
			expErr: "subnet mask must be longer",
		},
		{
			name:  "success IPv6 /98 cidr",
			cidr:  "2001:db8::/98",
			addrs: 1073741824,
		},
		{
			name:  "success IPv6 /66 cidr",
			cidr:  "2001:db8::/66",
			addrs: int64(1) << uint(62),
		},
		{
			name:   "failure unsupported IPv6 cidr",
			cidr:   "2001:db8::/65",
			expErr: "subnet mask must be longer",
		},
	}

	for _, tc := range testCases {
		_, cidr, err := net.ParseCIDR(tc.cidr)
		if err != nil {
			t.Fatal(err)
		}
		size, err := RangeSize(cidr)
		if err != nil {
			if tc.expErr == "" {
				t.Errorf("test %s failed, unexpected error: '%s'", tc.name, err.Error())
			} else if !strings.Contains(err.Error(), tc.expErr) {
				t.Errorf("test %s failed, expected '%s' error, got: '%s' error", tc.name, tc.expErr, err.Error())
			}
		} else {
			if tc.expErr == "" {
				if size != tc.addrs {
					t.Errorf("test %s failed. %s should have a range size of %d, got %d", tc.name, tc.cidr, tc.addrs, size)
				}
			} else {
				t.Errorf("test %s failed, expected error '%s' never occurred", tc.name, tc.expErr)
			}
		}
	}
}
```

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (danehans) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: I have updated the code to allow a /66 ipv6 cidr.

File: pkg/registry/core/service/ipallocator/allocator_test.go
Comment (danehans) [reply pkg/registry/core/service/ipallocator/allocator_test.go]: I have removed the v4 an v6 test cases that are redundant.

**Comment by danehans** [comment]
@bowei @mikedanese do you mind reviewing this PR when you have a moment?

**Comment by danehans** [comment]
@caesarxuchao do you mind reviewing this PR when you have a moment? This needed for adding IPv6 support in the 1.9 release.

/assign @thockin

File: pkg/registry/core/service/ipallocator/allocator_test.go
Comment (caesarxuchao) [inline pkg/registry/core/service/ipallocator/allocator_test.go]: The name `expect` isn't clear to me. What does it expect?

**Comment by caesarxuchao** [comment]
A nit. I don't see other problems, but I'm not experienced in the network world. @bowei @dnardo could you take a look?

/unassign @caesarxuchao

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (thockin) [inline pkg/registry/core/service/ipallocator/allocator.go]: I'm not sure what /1 means here - do you mean /31 ?  `10.0.0.0/1` is nonsense?

Comment (danehans) [reply pkg/registry/core/service/ipallocator/allocator.go]: I mean the mask must be /2 or longer (/8, /16, etc.). I did not change any of the existing IPv4 logic in this PR. I updated the error message because it was incorrect.

__Old Error:__`masks greater than 31 bits are not supported`

As you can see in the RangeSize function, `(bits - ones) >= 31`. For IPv4, bits = 32 and ones is the number of masked bits in the 32-bit address. For a /1, the math would trigger the error ``(32 - 1) >= 31`.  I am happy to submit a PR to implement a fix if this is not the expected behavior, unless you prefer the fix to be implemented in this PR.

**Comment by thockin** [comment]
I just mean that `/2` seems like the wrong notation - the "mask" is usually
the ones portion, right, so we're saying that masks subnets smaller than
/30 are not allowed.  And frankly, even /30 is kind of silly, but might be
useful for testing.  I am arguing with the words/messages, not the logic.

**Comment by danehans** [comment]
I can change the notation in the error message from a `/`, to specifying the number of masked bits supported. For example:

`ErrMaskTooSmall = errors.New("subnet mask must be longer than 1 bit for IPv4 or longer than 65 bits for IPv6")`

I used the `/` notation in the error message since a CIDR is given as parameter to the function, i.e. `10.10.0.0/16`. It's my [understanding](https://en.wikipedia.org/wiki/Subnetwork) that subnet masks are typically described using "length" as the number of masked bits (1's) in an address and I believe the [Size() method](https://golang.org/pkg/net/#IPMask.Size) follows this understanding.

Here is a Go Playground example: https://play.golang.org/p/6r4JwMs1TL

I apologize if I'm not following you.

**Comment by thockin** [comment]
I think there are a couple things being conflated.

The `panic()` test for > 31 is pretty horrible, I admit, and I don't actually recall why it is needed.  It certainly was not about policy of "too big" or "too small".  I don't see why it can't simply return 0, as in "the range you specified has 0 available addresses".  The policy of how to handle that should be at the callers, I think?

`NewAllocatorCIDRRange()` wants to subtract 2, so it should probably check that the values from `RangeSize()` is at least 2, otherwise return an error (or fake it like it does today).

Basically, could teh whole PR be reduced to removing that panic and just allowing it to return 0 ?

**Comment by danehans** [comment]
@thockin I like the idea of simply returning 0 and removing the error return. I can check `max >= 2` in `NewAllocatorCIDRRange`. If so, should `NewAllocatorCIDRRange` return an empty *range or add an error return to the function?

**Comment by danehans** [comment]
@thockin the PR has been greatly simplified and is passing tests.

**Comment by danehans** [comment]
@bowei when you have a moment, do you mind reviewing this PR? The PR has been refactored based on recent feedback from @thockin

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (pmichali) [inline pkg/registry/core/service/ipallocator/allocator.go]: As a future clean-up, we should look at making this easier to understand (or supplement it with a comment to indicate that if the network bits are less than 2 for IPv4 or less than 65 for IPv6 it fails. OK for now.

**Review by pmichali — APPROVED** [review]
/lgtm

File: pkg/registry/core/service/ipallocator/allocator.go
Comment (bowei) [reply pkg/registry/core/service/ipallocator/allocator.go]: Looks like it is not too much work to propagate an error instead of panic here. Would be great to fix.

Comment (danehans) [reply pkg/registry/core/service/ipallocator/allocator.go]: @bowei I removed the panic. @thockin asked to return 0 instead of an error/panic.

**Comment by danehans** [comment]
@thockin this PR needs one more approver. Do you mind taking a look when you have a few minutes to spare?

**Comment by bowei** [comment]
@wojtek-t for approval

**Pull request merged**
